Dockerfile

```diff
@@ -28,14 +28,30 @@ FROM golang:1.7.6
 #ENV https_proxy ""
 ARG http_proxy
 ARG https_proxy
-ENV GOPATH=/go/ NET_CONTAINER_BUILD=1
+
+WORKDIR /go/src/github.com/contiv/netplugin/
 
 ENTRYPOINT ["netplugin"]
 CMD ["--help"]
 
-COPY ./ /go/src/github.com/contiv/netplugin/
+# by far, most of the compilation time is building vendor packages
+# build the vendor dependencies as a separate docker caching layer
+COPY ./vendor/ /go/src/github.com/contiv/netplugin/vendor/
+# there is a bug in go-winio, remote the grep -v after this merges:
+# https://github.com/contiv/netplugin/pull/999
+RUN GOGC=1500 \
+  go install -ldflags "-s -w" \
+             $(go list ./vendor/... | grep -v go-winio)
 
-WORKDIR /go/src/github.com/contiv/netplugin/
+# build the netplugin binaries
+COPY ./ /go/src/github.com/contiv/netplugin/
 
-RUN make build
+ARG BUILD_VERSION=""
+ARG USE_RELEASE=""
 
+RUN GOPATH=/go/ \
+    BUILD_VERSION="${BUILD_VERSION}" \
+    USE_RELEASE="${USE_RELEASE}" \
+    make compile \
```

Comment:
If you are compiling netplugin as part of docker build then you should add atleast checks here. Otherwise, we will end up not catching errors. If the idea is to make a quick build knowing that there are no code changes (for releases), then Dockerfile should only compile vendor and build the container. User can decide to skip tests when running the container using docker exec

Reply:
I have checks for docker based build as a backlog item but this isn't replacing other ways to build so it's not required (yet), and this PR is long enough already.

I was planning to have a separate Dockerfile that does the checks after compile which I think matches development flow better and will be faster to iterate.  Also, will allow choice as to when to run the checks based on which target is being run in the Makefile.

Reply:
Please add the development Dockerfile here (root directory) and move the one with no checks to a release directory. Please take care of it in a later PR